compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -225,17 +225,31 @@ public BType checkType(Location pos,
     }
 
     public boolean isJSONContext(BType type) {
-        if (type.tag == TypeTags.UNION) {
-            return ((BUnionType) type).getMemberTypes().stream().anyMatch(memType -> memType.tag == TypeTags.JSON);
+        switch (type.tag) {
+            case TypeTags.MAP:
+                return isJSONContext(((BMapType) type).constraint);
+            case TypeTags.ARRAY:
+                return isJSONContext(((BArrayType) type).eType);
```

Comment:
Shouldn't this only be applicable for `map<json>` and JSON arrays? For example, with this check, types such as `map<json|Foo>` (where `Foo` is an object) would also be considered as a JSON context right?

Reply:
I think we should clarify what JSON context means. I guess one could argue that even `map<json|Foo>` can be considered a JSON context since it still allows JSON, and similarly something like `map<int|Foo>`.

Reply:
Refer https://github.com/ballerina-platform/ballerina-lang/issues/28342#issuecomment-806403455 for clarification regarding the JSON-related context

Reply:
My concern is not cases where we add elements to maps such as `map<json|Foo>`. I think those count as JSON context. But the above makes the following case to be considered as a JSON context too right? 
```ballerina
map<json|Foo>? m = null;
```

Reply:
If we consider `map<json|Foo>` as a JSON context isn't this similar to the original example `int?`?

Reply:
Ah now only I saw that @lasinicl has already raised this in the issue. So according to that, use of `null` anywhere is allowed by the compiler I guess? 🤔

Reply:
^ Ah nvm my bad, ignore the above. Really should start checking before commenting 😅